### PR TITLE
feat: Migrate repository to TypeScript

### PR DESCRIPTION
**Pull request by jimmycallin**
- [ ] I have added automatic tests where applicable
- [x] The PR title is suitable as a release note
- [x] The PR contains a description of what has been changed
- [x] The description contains manual test instructions

## Changes

- Migrated ftrack-web-widget to TypeScript
- Uses same `tsconfig.json` as `yarn create vite` outputs when choosing a TypeScript project.
- Updates jsdoc comments to tsdoc.

## Test

- Check that it builds
- Check that `yarn lint` outputs typescript errors 
- Check that the config looks good
- Check that it outputs d.ts files and that they are picked up when used in other projects.

### CODE REVIEW

**Review by WaylonFtrack — CHANGES_REQUESTED** [review]

File: src/ftrackWidget.ts [hunk 1]
```diff
@@ -3,13 +3,40 @@
...
  * Handle communication with the ftrack web application.
  */
-let targetOrigin = null;
-let credentials = null;
-let entity = null;
-let onWidgetLoadCallback, onWidgetUpdateCallback;
 
-/** Open sidebar for *entityType*, *entityId*. */
-export function openSidebar(entityType, entityId) {
+type MessageContent = {
```
[the remark this answers is not in the document]
Comment (jimmycallin) [reply src/ftrackWidget.ts]: interesting, thanks!

File: .eslintrc [hunk 2]
```diff
@@ -1,3 +1,6 @@
 {
-  "extends": "react-app"
+  "extends": ["react-app", "plugin:@typescript-eslint/recommended"],
```
[the remark this answers is not in the document]
Comment (jimmycallin) [reply .eslintrc:2]: we use eslint-config-react-app for linting in all our projects, even when they are not containing react as dependency. it's not ideal but preferable over having separate eslint configs to maintain.

**Comment by jimmycallin** [comment]
@WaylonFtrack please review again!

File: src/ftrackWidget.ts [hunk 3]
```diff
@@ -185,12 +229,12 @@ function onDocumentKeyDown(event) {
     "charCode",
     "keyCode",
     "which",
-  ];
-  const eventData = {};
-  for (let i = 0; i < fields.length; i += 1) {
-    const field = fields[i];
-    eventData[field] = event[field];
-  }
+  ] as (keyof KeyboardEvent)[];
```
[the remark this answers is not in the document]
Comment (WaylonFtrack) [reply src/ftrackWidget.ts]: This writing feels a bit strange, maybe `Array<keyof KeyboardEvent>` is better?

File: src/ftrackWidget.ts [hunk 4]
```diff
@@ -3,13 +3,41 @@
...
  * Handle communication with the ftrack web application.
  */
-let targetOrigin = null;
-let credentials = null;
-let entity = null;
-let onWidgetLoadCallback, onWidgetUpdateCallback;
 
-/** Open sidebar for *entityType*, *entityId*. */
-export function openSidebar(entityType, entityId) {
+interface MessageContent {
```
Comment (WaylonFtrack) [inline src/ftrackWidget.ts:7]: these interface should be exported, it may be used by users.

**Review by WaylonFtrack — CHANGES_REQUESTED** [review]

File: src/ftrackWidget.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (WaylonFtrack) [reply src/ftrackWidget.ts]: This writing feels a bit strange, maybe `Array<keyof KeyboardEvent>` is better?

File: src/ftrackWidget.ts [hunk 5]
```diff
@@ -161,18 +207,19 @@ function onDocumentClick(event) {
...
-    event.target.isContentEditable
+    target.isContentEditable
   ) {
-    return true;
+    return;
   }
 
   // Copy event data to KeyboardEvent constructor argument.
-  const fields = [
+  const fields: (keyof KeyboardEvent)[] = [
```
Comment (WaylonFtrack) [inline src/ftrackWidget.ts:222]: This writing feels a bit strange, maybe `Array<keyof KeyboardEvent>` is better?

**Comment by jimmycallin** [comment]
@WaylonFtrack thanks! please review again :)

**Review by CzBiX — APPROVED** [review]
LGTM

**Review by WaylonFtrack — APPROVED** [review]

**Review by gismya — CHANGES_REQUESTED** [review]
Very good! Just a few small comments.

I haven't checked if it actually works after the refactor though - I don't know enough about the package to do that.

File: src/ftrackWidget.ts [hunk 6]
```diff
@@ -3,13 +3,41 @@
...
+let targetOrigin: string;
+let credentials: { serverUrl: string };
+let entity: EntityType;
+let onWidgetLoadCallback: (content: MessageContent) => void,
+  onWidgetUpdateCallback: (content: MessageContent) => void;
+
+/**
+ * Open sidebar for *entityType*, *entityId*.
+ */
+export function openSidebar(entityType: string, entityId: string) {
```
[the remark this answers is not in the document]
Comment (jimmycallin) [reply src/ftrackWidget.ts:40]: don't think it hurts!

File: src/ftrackWidget.ts [hunk 7]
```diff
@@ -161,18 +207,21 @@ function onDocumentClick(event) {
...
  */
-function onDocumentKeyDown(event) {
+function onDocumentKeyDown(event: KeyboardEvent) {
   // Ignore events when focus is in an textarea/input/contenteditable.
-  const tagName = event.target.tagName.toLowerCase();
+  const target = event.target as HTMLElement | null;
+  if (!target) return;
+
+  const tagName = target?.tagName.toLowerCase();
   if (
```
[the remark this answers is not in the document]
Comment (jimmycallin) [reply src/ftrackWidget.ts:216]: no opinion from my part - kept it as it was in the implementation before, feels out of scope to make any changes outside required for typescript migration.

**Comment by jimmycallin** [comment]
@gismya addressed your comments, please review again!

File: src/ftrackWidget.ts [hunk 8]
```diff
@@ -209,8 +212,8 @@ function onDocumentClick() {
  */
 function onDocumentKeyDown(event: KeyboardEvent) {
   // Ignore events when focus is in an textarea/input/contenteditable.
-  const target = event.target as HTMLElement | null;
-  if (!target) return;
+  const target = event.target;
+  if (!target || !(target instanceof HTMLElement)) return;
 
   const tagName = target?.tagName.toLowerCase();
```
Comment (gismya) [inline src/ftrackWidget.ts:218]: The optional chaining after target is still unnecessary.

File: src/ftrackWidget.ts [hunk 9]
```diff
@@ -67,8 +100,10 @@ export function openPreview(componentId) {
   );
 }
 
-/** Navigate web app to *entityType*, *entityId*. */
-export function navigate(entityType, entityId, module) {
+/**
+ * Navigate web app to *entityType*, *entityId*.
+ */
+export function navigate(entityType: string, entityId: string, module: string) {
```
Comment (gismya) [inline src/ftrackWidget.ts:106]: Same change as in openSidebar possible here, if we decide on going that route.

File: src/ftrackWidget.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (gismya) [reply src/ftrackWidget.ts]: But does it help? Can't find much discussion online about it, and feels presumptuous to just go with my gut feeling on this (Seeing as it in some sense complicates the code)

**Review by gismya — CHANGES_REQUESTED** [review]
Just want to land in whether we use the indexed access types for the functions. If we do that, also change navigate function, if not, roll back openSidebar function. Would love to hear if @WaylonFtrack has any thoughts on if that pattern makes sense.

File: src/ftrackWidget.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (gismya) [reply src/ftrackWidget.ts]: image

File: src/ftrackWidget.ts [hunk 10]
```diff
@@ -161,18 +210,21 @@ function onDocumentClick(event) {
 /**
  * On document keydown forward to parent application.
  */
-function onDocumentKeyDown(event) {
+function onDocumentKeyDown(event: KeyboardEvent) {
   // Ignore events when focus is in an textarea/input/contenteditable.
-  const tagName = event.target.tagName.toLowerCase();
+  const target = event.target;
+  if (!target || !(target instanceof HTMLElement)) return;
```
Comment (gismya) [inline src/ftrackWidget.ts:216]: I also realised that `!(target instanceof HTMLElement)` is an implicit null check as well, but I feel it's probably more readable doing that null check "twice".

File: src/ftrackWidget.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (CzBiX) [reply src/ftrackWidget.ts]: If we follow the textbook approach, then we should use the indexed type approach. This is particularly helpful for refactoring scenarios, such as `entityType.id` going from a `number` to a `string`.
But if this field is primitive type, I personally prefer to just use the primitive type, which is simpler to write and understand.

[the remark this answers is not in the document]
Comment (WaylonFtrack) [reply src/ftrackWidget.ts]: If we follow the textbook approach, then we should use the indexed type approach. This is particularly helpful for refactoring scenarios, such as entity.id going from a number to a string.
But if this field is primitive type, I personally prefer to just use the primitive type, which is simpler to write and understand.

**Comment by jimmycallin** [comment]
rolled back indexed type after weiran's feedback and removed the null check, please review again!

**Review by gismya — APPROVED** [review]

**Review by emm4back — APPROVED** [review]

**Review by swetapadmanabhan — APPROVED** [review]

**Review by mertkirbuga — APPROVED** [review]

File: src/ftrackWidget.ts [hunk 11]
```diff
@@ -186,11 +235,10 @@ function onDocumentKeyDown(event) {
     "keyCode",
     "which",
   ];
-  const eventData = {};
-  for (let i = 0; i < fields.length; i += 1) {
-    const field = fields[i];
-    eventData[field] = event[field];
-  }
+
+  const eventData: Partial<KeyboardEvent> = Object.fromEntries(
```
Comment (lucaas) [inline src/ftrackWidget.ts:239]: Doesn't looks like this get transpiled, so we are raising the requirement to browsers supporting [Object.fromEntries](https://caniuse.com/?search=fromEntries) and [arrow functions](https://caniuse.com/?search=arrow%20function).

Personally, I would have kept the for-loop, but I think the browser support is enough for this to be ok.

**Review by lucaas — APPROVED** [review]
Tested and working well.

Added a few optional comments to the code. Read through before you merge.

File: src/ftrackWidget.ts [hunk 12]
```diff
@@ -42,7 +68,9 @@ export function openActions(selection) {
   );
 }
 
-/** Close action window for current widget. */
+/**
```
[the remark this answers is not in the document]
Comment (jimmycallin) [reply src/ftrackWidget.ts:79]: Not sure, I think I changed it mostly for consistency but don't really have a strong opinion. I'll keep it as is and won't change them in the future :)

**Pull request merged**